File: 2170-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.cpp
class Solution {
public:
    int f(int i,int n,vector<int>& nums,int curr,int mx, vector<vector<int>>& dp)
    {
        if(i==n)
        return curr==mx;
        if(dp[i][curr]!=-1)
        return dp[i][curr];
        int nt=f(i+1,n,nums,curr,mx,dp);
        int t=f(i+1,n,nums,curr|nums[i],mx,dp);
        return dp[i][curr]=t+nt;
    }
    int countMaxOrSubsets(vector<int>& nums) {
        int mx=0;
        int n=nums.size();
        for(int i=0;i<n;i++)
            mx|=nums[i];
        vector<vector<int>>dp(n,vector<int>(mx+1,-1));
        return f(0,n,nums,0,mx,dp);
    }
};

File: 2170-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.py
class Solution:
    def countMaxOrSubsets(self, nums: List[int]) -> int:
        n = len(nums)
        mx = 0
        for num in nums:
            mx |= num

        @lru_cache(None)
        def f(i: int, curr: int) -> int:
            if i == n:
                return 1 if curr == mx else 0
            nt = f(i + 1, curr)
            t = f(i + 1, curr | nums[i])
            return t + nt

        return f(0, 0)
